### ISSUE TITLE
Function argument hints and docs only shown with an additional comma

### ISSUE THREAD

**Issue by NiklasRosenstein** [microsoft/pylance-release#1128]
## Environment data

-   Language Server version: 2021.3.4
-   OS and version: osx 15.15.7
-   Python version (& distribution if applicable, e.g. Anaconda): Miniconda Python 3.9.1

## Expected behaviour

Inside a function call, after a comma, the language server shows hints about the function arguments and it's docstrings.

## Actual behaviour

Pylance only shows that information after an additional comma.

![Screen Shot 2021-04-05 at 12 46 51](https://user-images.githubusercontent.com/1318438/113623913-49412e00-965f-11eb-8a94-e85ab73beb9e.png)
![Screen Shot 2021-04-05 at 12 47 05](https://user-images.githubusercontent.com/1318438/113623918-4ba38800-965f-11eb-88e9-b08894720fad.png)

## Logs

```
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: getSemanticTokens delta
[BG(1)] getSemanticTokens delta previousResultId:1617655032602 at /Users/nrosenstein/projects/craftr/craftr-dsl/src/craftr/dsl/__init__.py ...
[BG(1)]   parsing: /Users/nrosenstein/projects/craftr/craftr-dsl/src/craftr/dsl/__init__.py (3ms)
[BG(1)]   binding: /Users/nrosenstein/projects/craftr/craftr-dsl/src/craftr/dsl/__init__.py (0ms)
[BG(1)] getSemanticTokens delta previousResultId:1617655032602 at /Users/nrosenstein/projects/craftr/craftr-dsl/src/craftr/dsl/__init__.py (28ms)
Background analysis message: analyze
[BG(1)] analyzing: /Users/nrosenstein/projects/craftr/craftr-dsl/src/craftr/dsl/__init__.py ...
[BG(1)]   checking: /Users/nrosenstein/projects/craftr/craftr-dsl/src/craftr/dsl/__init__.py (3ms)
[BG(1)] analyzing: /Users/nrosenstein/projects/craftr/craftr-dsl/src/craftr/dsl/__init__.py (3ms)
Background analysis message: resumeAnalysis
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: setFileOpened
Background analysis message: markFilesDirty
[FG] parsing: /Users/nrosenstein/projects/craftr/craftr-dsl/src/craftr/dsl/__init__.py (3ms)
[FG] binding: /Users/nrosenstein/projects/craftr/craftr-dsl/src/craftr/dsl/__init__.py (1ms)
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
Background analysis message: getSemanticTokens delta
[BG(1)] getSemanticTokens delta previousResultId:1617655035941 at /Users/nrosenstein/projects/craftr/craftr-dsl/src/craftr/dsl/__init__.py ...
[BG(1)]   parsing: /Users/nrosenstein/projects/craftr/craftr-dsl/src/craftr/dsl/__init__.py (2ms)
[BG(1)]   binding: /Users/nrosenstein/projects/craftr/craftr-dsl/src/craftr/dsl/__init__.py (0ms)
[BG(1)] getSemanticTokens delta previousResultId:1617655035941 at /Users/nrosenstein/projects/craftr/craftr-dsl/src/craftr/dsl/__init__.py (26ms)
Background analysis message: analyze
[BG(1)] analyzing: /Users/nrosenstein/projects/craftr/craftr-dsl/src/craftr/dsl/__init__.py ...
[BG(1)]   checking: /Users/nrosenstein/projects/craftr/craftr-dsl/src/craftr/dsl/__init__.py (1ms)
[BG(1)] analyzing: /Users/nrosenstein/projects/craftr/craftr-dsl/src/craftr/dsl/__init__.py (1ms)
Background analysis message: resumeAnalysis
Background analysis message: getDiagnosticsForRange
Background analysis message: getDiagnosticsForRange
```

**Comment by jakebailey** [comment]
Is there a reason why your call and dict literal are unclosed? VS Code defaults to ensuring matching parens/braces/brackets.

Our parser tries to recover as best as it can, but without the ending paren and curly brace, we might be getting things wrong. Can you add in the `)` and `}` and retest?

**Comment by NiklasRosenstein** [comment]
Hey @jakebailey!

I have that feature disabled in VSCode because I prefer to type the closing brackets and quotation myself. 

Sure no problem. Unfortunately it looks like that doesn't fix it.

![image](https://user-images.githubusercontent.com/1318438/113625026-a5f11880-9660-11eb-9a70-f6787330aac8.png)
![image](https://user-images.githubusercontent.com/1318438/113625038-abe6f980-9660-11eb-9c34-7f001218eb51.png)

Seems to me like something must be wrong here to begin with, given that the double comma is actually invalid syntax and the single comma is valid (with the closing parentheses).

Cheers!

**Comment by jakebailey** [comment]
Thanks. Is the code for `Closure` available somewhere?

**Comment by NiklasRosenstein** [comment]
Actually what is curious is that with the closing control characters, it emphasises the right argument (here: owner), and without them it emphasises the argument after (i.e. delegate). Also some other things seems to be happening here when moving the cursor.

So in all of the below cases, the two commas are necessary for the hint to appear at all. But only in the case where the closing control characters are present, _and_ the cursor is located after the second comma, it emphasises the correct third argument.

![image](https://user-images.githubusercontent.com/1318438/113625365-231c8d80-9661-11eb-94e1-9185635f77f6.png)

![image](https://user-images.githubusercontent.com/1318438/113625386-2adc3200-9661-11eb-9cf4-c09e7c8b76b8.png)

![image](https://user-images.githubusercontent.com/1318438/113625418-33cd0380-9661-11eb-909c-0387a3a386e7.png)

Yes, you can find the code to that class here:

https://github.com/craftr-build/craftr/blob/5.x/craftr-core/src/craftr/core/closure/__init__.py

**Comment by erictraut** [comment]
It appears that the signature help is dismissed and doesn't reappear when a nested call expression is completed.

Here's a simplified repro:
```python
def func1(a: int, b: int) -> None: ...
def func2() -> int: ...

func1(func2(),
```

**Comment by erictraut** [comment]
I found the bug. We use a heuristic in the signature provider to deal with "broken" parse trees, such as when the expression is missing a closing parenthesis. This heuristic was failing in this particular case. I've added some additional logic to prevent this failure.

@jakebailey, here's a PR for you to review when you have time.

**Comment by jakebailey** [comment]
This issue has been fixed in version 2021.4.0, which we've just released. You can find the changelog here: https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202140-7-april-2021

**Comment by NiklasRosenstein** [comment]
I can confirm it works. Thank you! 😃
